### DISCUSSION TITLE
Extracting image annotations

### DISCUSSION THREAD

**Discussion by 2246sam20** [ImageMagick/ImageMagick#7582] [Help] [answered]
### ImageMagick version

7.1.1-37

### Operating system, version and so on

Windows 11

### Description

Hello! Hope you are doing well.
I have a query it is as follows
1. Lets say i have some drawings(DG1) on a canvas and then i overlay an image(IM) over those drawings
2. Then I annotate(AN) something  on the image overlayed over the canvas 
3. So, can I extract the image annotations and save them in an image file by excluding the drawings which were below the image?
4. To make things clear the z order of the elements is as follows in a bottom to top manner
    DG1 -> IM -> AN
 
Thanks in advance!

**Comment by snibgo** [comment]
Do you know what the canvas looked like before doing the annotation? Was that image fully opaque? Is the annotation in a different colour to what was on the canvas?

If the answer to these is "yes", then you can extract the annotation. A simple difference between the canvas before and after the annotation shows what pixels have changed. If the annotation was fully opaque, then the canvas pixel colours after the annotation are the annotation. If the annotation included partial transparency, then if the annotation was a single colour (with varying transparency) then the transparency can be calculated.

**Reply by 2246sam20** [reply]
Hello! Thanks for the reply
To answer your first question, the scenario is like lets say I have some data points(basically array of {x,y} coordinates ) of graphical elements and some image elements in a page. I am drawing these different elements and saving them to a png file using Image Magick.
So to check if the graphical element was drawn over the image( which will make it an annotation), I have to check each point if it lies within the image bounds, which is a very inefficient approach.
So in a nutshell I cant know for sure if the element is an annotation without doing the the aforementioned calculation.
So, I was wondering if I could extract the annotations of the image using Image Magick after rendering all the elements and saving them into a png file, without checking each point.
Answer to your second question - yes, images are fully opaque.

**Comment by snibgo** [comment]
To help me understand your problem, please show examples.

**Reply by 2246sam20** [reply]
Hello! Hope you are doing fine 
I have attached the files for a better understanding

Annotations of respective images -> 
![1_annotation](https://github.com/user-attachments/assets/8bbe7ea3-84f1-46e0-94f8-215e1dc20f9f)
![2_annotation](https://github.com/user-attachments/assets/608d6fb9-24bd-437a-9220-a77c0e7db9f7)
![3_annotation](https://github.com/user-attachments/assets/f503086c-317b-4400-acd2-c08a612b0c30)

merged.jpeg -> ![merged](https://github.com/user-attachments/assets/233b4c64-ac15-4318-929a-fe2a43c4b4bd)

page.jpeg -> ![page](https://github.com/user-attachments/assets/55f6a681-42f8-42cb-bf19-6b1f6437e8f0)


1. merged -> It contains all the drawings, images, and annotations on the images.
2. page -> It contains all the drawings without images and their annotations.
3. <image_no>_annotation -> contains annotations of respective images.

Currently, I am obtaining  mergd.jpeg and page.jpeg files and  I have rendered graphical elements on them using ImageMagick.
So I want to extract the annotations of each image using some image processing, without checking each point of the graphical elements if it lies within the image bounds.


In a nutshell I want something like this 

`(merged.jpeg, page.jpeg) -> Image Processing -> annotations.jpeg
`
P.S.- Also It can be possible that below the images in the merged.jpeg there may already be some drawings. And they are not be considered as annotations and they will be part of the page.jpeg file 

P.S.S -  Sorry for the late reply.

**Comment by fmw42** [comment]
If you know the locations and have the graphical images that you inserted and want to see if they are there, you could crop the image at the location for each of the graphical image and do a -compare difference -composite and then get the mean or std to test if it is there.

**Reply by 2246sam20** [reply]
Hello @fmw42 
Please refer to this reply of mine  https://github.com/ImageMagick/ImageMagick/discussions/7582#discussioncomment-10688319

**Comment by fmw42** [comment]
I do not understand. The final annotations image does not correspond to your 3 annotations images. And the 3 annotations are at a different scale to what is in the merged image. And you do show the graphical elements (images) that went into the merged image.  Please clarify.

Also clarify your exact input image(s) that are available in your actual processing problem and show what you expect for the output.

**Reply by 2246sam20** [reply]
hello!
I hope this reply will make things more clear-> https://github.com/ImageMagick/ImageMagick/discussions/7582#discussioncomment-10707935

**Comment by snibgo** [comment]
I think your problem is: given images merged.jpeg and page.jpeg, derive one image that contains the grey background from page.jpeg and the yellow marks that are in merged.jpeg but not in page.jpeg. The output would not contain the added images that are in merged.jpeg but not in page.jpeg.

Is that what you want?

That isn't possible. The reason is that the area around the "2" in merged.jpg has no corresponding pixels in page.jpeg. So we can't get a complete gray background. Instead, I will assume you want a transparent background. So the result will contain yellow pixels (from the yellow marks) on a transparent background.

Your images are jpeg. This is lossy. You should never do image processing using inputs that have been through jpeg compression, because the processing is much harder.

1. The first problem is to align merged.jpeg and page.jpeg. They are different scales. page.jpeg has extra pixels on the left. merged.jpeg has extra pixels on the right and bottom. So a process would find the area that is common to both, and the resize that would make the yellow marks align. This task is not trivial.

2. The yellow marks are anti-aliased. This means that pixels at the edges of the yellow marks are a mixture of two colours: the yellow mixed with the background that might be gray or blue or orange. So isolating the yellow marks is subjective: there is no objective rule that determines which pixels are yellow marks. In merged.jpg, near the top of the "2" there are yellowish pixels that appear to be part of the yellow mark, but in the second annotated image those pixels are gray background.

Assuming problems (1) and (2) are solved, then we have two images. One contains just the yellow pixels from merged.jpeg, and the other contains just the yellow pixels from page.jpeg. So we can then find the yellow pixels that are in just the first of these resuts. The task is done.

**Reply by 2246sam20** [reply]
Hello! Thanks for the reply, I really appreciate it.
Sorry for the confusion, let me make the things more clear.
-> Dimensions of merged and page jpeg -> 3840*2160.
-> Image Format - PNG.
-> Background is transparent.
-> In the merged image drawing is in yellow color and the  brush color and width is always known .
Also the photos I attached are the ones I took from my phone camera so those were not the original ones.
And I cropped the annotations image to show just the drawing part but it created more confusion.
So the problem is as follows:
From the merged image I want to extract the yellow drawings only that are over the image elements.  
-> The coordinates of the three images in the merged file are known.

![merged](https://github.com/user-attachments/assets/157e86d9-40a5-4288-8e87-bd56a6cc2cbe)

In the above merged file  I have circled the image elements in the red, and i want to extract the yellow drawings that are over the image elements only.
After extracting the annotations it will be saved in PNG file with a transparent background and with dimensions similar to the merged and page image files.
Dont get confused with the jpeg files I attached as they are not the original ones  I just took them from my camera and cropped them.

I hope things are clear now than before.

**Comment by snibgo** [comment]
> -> Dimensions of merged and page jpeg -> 3840*2160.
-> Image Format - PNG.
-> Background is transparent.

None of that is true of the images you posted.

> Also the photos I attached are the ones I took from my phone camera so those were not the original ones.
... they are not the original ones I just took them from my camera and cropped them.

Huh? What? You displayed some images on your screen, and photographed those, and posted those photos? Why did you do that? Why not post the original images?

**Reply by 2246sam20** [reply]
Yes, I cant get the original ones for some reasons. 
That’s why i clicked the pictures of them.
Just to make the problem clear , I uploaded the clicked pictures.
But I hope the problem is clear now.

**Comment by snibgo** [comment]
Suppose we have two input images, imageA.png and imageB.png. Both images have transparency.

imageA.png is:

![imageA](https://github.com/user-attachments/assets/a4734c72-db9f-48da-ae5a-b17a4fe26636)

imageB.png is:

![imageB](https://github.com/user-attachments/assets/36314f69-d887-4803-a029-3f76baad4d02)

Suppose we want an output that is yellow where imageA is yellow and imageB is *not* yellow. Elsewhere, the output should be transparent. Windows BAT script:
```
magick ^
  imageA.png imageB.png ^
  ( -clone 0 +transparent Yellow -alpha extract ) ^
  ( -clone 1 +transparent Yellow -alpha extract -negate ) ^
  ( -clone 2-3 -alpha off -compose Darken -composite ) ^
  -delete 1-3 ^
  -alpha off -compose CopyOpacity -composite ^
  out.png
```
The result out.png is:

![out](https://github.com/user-attachments/assets/41425af4-5f64-4937-a46c-63e2a7ee94fe)

Does that help?

**Reply by 2246sam20** [reply]
Hello, Thanks for bearing with me, I really appreciate the help.

> Suppose we want an output that is yellow where imageA is yellow and imageB is not yellow. Elsewhere, the output should be transparent.

Thats almost what I am looking for.
But, Suppose imageB is like this

![imageB](https://github.com/user-attachments/assets/a6b54e0a-f0ae-4a5c-88b9-4ac48122c16f)

Then still can we get the output image as out.png which you attached above?

**Comment by snibgo** [comment]
You can try the command I showed you with that image for yourself. The result is, of course, identical to the previous result.

**Reply by 2246sam20** [reply]
Okay!
Thanks for your support i will check it.

**Reply by 2246sam20** [reply]
Thanks you so much @snibgo, its working. Really appreciate the help.
I have just one more doubt :P.
what if we have more than one color?  Do we have to run the above bat file for every color? Or can we optimize this somehow?

**Comment by snibgo** [comment]
It depends on what the processing you want. I supposed the processing is "we want an output that is yellow where imageA is yellow and imageB is not yellow. Elsewhere, the output should be transparent." If you have more than one colour, what processing do you want?

**Reply by 2246sam20** [reply]
I want to extract the output just like the above for every color. I think I need to do the processing for each color.
Or we can extract  all the colors in a one go?
Lets assume that I have a list that contains all the colors.

**Comment by snibgo** [comment, marked as answer]
> I want to extract the output just like the above for every color. I think I need to do the processing for each color.

This is vague. Perhaps you want:

Each output pixel should be transparent, except in the following circumstances:
1. If imageA is yellow and imageB is not yellow, make the output yellow.
2. If imageA is red and imageB is not red, make the output red.
3. Repeat the previous for any other colours.

Suppose imageA3.png is:

![imageA3](https://github.com/user-attachments/assets/c040e007-75f6-49ac-b8ae-a7944ddfd134)

And imageB3.png is:

![imageB3](https://github.com/user-attachments/assets/c5af19fb-9ace-4461-8b8f-3279517c02da)

```
magick ^
  imageA3.png -write mpr:InA +delete ^
  imageB3.png -write mpr:InB +delete ^
  ( mpr:InA +transparent Yellow -alpha extract ) ^
  ( mpr:InB +transparent Yellow -alpha extract -negate ) ^
  ( -clone -1,-2 -alpha off -compose Darken -composite ) ^
  -delete -2,-3 ^
  ( mpr:InA +transparent Red -alpha extract ) ^
  ( mpr:InB +transparent Red -alpha extract -negate ) ^
  ( -clone -1,-2 -alpha off -compose Darken -composite ) ^
  -delete -2,-3 ^
  ( -clone 0,-1 ^
    -compose Lighten -composite ^
  ) ^
  -delete 0--2 ^
  mpr:InA +swap ^
  -alpha off -compose CopyOpacity -composite ^
  out3.png
```

The result out3.png is:

![out3](https://github.com/user-attachments/assets/cd2bd4c7-c53a-4050-bb3a-93782059a23e)
Does this do what you want?

There is one `magick` command, which reads each input once.

You can see that in the command, three lines repeat for each colour. Suppose you have a text file that contains colours, one line per colour. Each colour might be `yellow` or `#ff0` etc. Then a shell script (eg BAT file) could read that text file and 
create a `magick` command, then run that command. With many colours this might break the limit on shell commend length, so instead the shell script creates and runs an ImageMagick script.

Here is the shell (Windows BAT) script, `yellowLine.bat`:
```
del cols.scr 2>nul:
echo off
(
  for /F %%C in (cols.txt) do (
    echo ^( mpr:InA +transparent %%C -alpha extract ^)
    echo ^( mpr:InB +transparent %%C -alpha extract -negate ^)
    echo ^( -clone "-1,-2" -alpha off -compose Darken -composite ^)
    echo -delete "-2,-3"
  )
  echo ^( -clone 0,-1
  echo    -compose Lighten -composite
  echo ^)
  echo -delete 0--2
  echo mpr:InA +swap
  echo -alpha off -compose CopyOpacity -composite
  echo -write out4.png
  echo -exit
) >cols.scr
echo on

magick ^
  imageA3.png -write mpr:InA +delete ^
  imageB3.png -write mpr:InB +delete ^
  -script cols.scr
```
When we run `yellowLine.bat`, the script creates and runs the IM script `cols.scr`, and creates `out4.png`, which is identical to `out3.png` above.

I haven't tested this with more than two colours.

Incidentally, the IM `cols.scr` that is created and run is:
```
( mpr:InA +transparent Yellow -alpha extract )
( mpr:InB +transparent Yellow -alpha extract -negate )
( -clone "-1,-2" -alpha off -compose Darken -composite )
-delete "-2,-3"
( mpr:InA +transparent \#f00 -alpha extract )
( mpr:InB +transparent \#f00 -alpha extract -negate )
( -clone "-1,-2" -alpha off -compose Darken -composite )
-delete "-2,-3"
( -clone 0,-1
   -compose Lighten -composite
)
-delete 0--2
mpr:InA +swap
-alpha off -compose CopyOpacity -composite
-write out4.png
-exit
```

**Reply by 2246sam20** [reply]
Cool.. thanks for the help.

**Reply by 2246sam20** [reply]
Hello!
I tried to write the C++ program version of the above commands you gave.
Can you please check it once since its throwing error at this line from the below code

```
Image combinedMask;
        combinedMask.composite(maskA, 0, 0, DarkenCompositeOp);

```

> The error is no pixels defined in cache.


```
void ImageProcessingUtil::extractAnnotations(const std::string& inputPathA, const std::string& inputPathB,
                                             const std::string& outputPath, const std::vector<Color>& colors) {
    InitializeMagick(nullptr);

    try {
        logMessage("Loading input images...");
        Image imageA(inputPathA);
        Image imageB(inputPathB);

        logMessage("Creating color masks...");
        auto combinedMasks = createColorMasks(imageA, imageB, colors);

        logMessage("Combining masks...");
        Image finalMask = combineMasks(combinedMasks);

        logMessage("Applying mask to image...");
        applyMaskToImage(imageA, finalMask);

        logMessage("Saving output image...");
        saveImage(imageA, outputPath);

        logMessage("Image processing completed successfully.");
    }
    catch (Exception& error) {
        std::ostringstream errorMsg;
        errorMsg << "Error processing images: " << error.what();
        logError(errorMsg.str());
    }
}

Image ImageProcessingUtil::createMaskForColor(const Image& image, const Color& color) {
    Image mask(image);
    mask.transparent(color);
    mask.alphaChannel(ExtractAlphaChannel);
    return mask;
}

std::vector<Image> ImageProcessingUtil::createColorMasks(const Image& imageA, const Image& imageB,
                                                        const std::vector<Color>& colors) {
    std::vector<Image> combinedMasks;

    for (const auto& color : colors) {
        std::ostringstream colorMsg;
        colorMsg << "Processing color: " << color;
        logMessage(colorMsg.str());

        Image maskA = createMaskForColor(image
A, color);
        Image maskB = createMaskForColor(imageB, color);
        maskB.negate();

       Image combinedMask;
        combinedMask.composite(maskA, 0, 0, DarkenCompositeOp);
        combinedMask.composite(maskB, 0, 0, DarkenCompositeOp);

        combinedMasks.push_back(combinedMask);
    }

    return combinedMasks;
}

Image ImageProcessingUtil::combineMasks(const std::vector<Image>& masks) {
    Image finalMask = masks[0];

    for (size_t i = 1; i < masks.size(); ++i) {
        finalMask.composite(masks[i], 0, 0, LightenCompositeOp);
    }

    return finalMask;
}

void ImageProcessingUtil::applyMaskToImage(Image& image, const Image& mask) {
    image.composite(mask, 0, 0, CopyAlphaCompositeOp);
}

void ImageProcessingUtil::saveImage(const Image& image, const std::string& outputPath) {
    image.write(outputPath);
}
```

**Reply by snibgo** [reply]
>Image combinedMask;
        combinedMask.composite(maskA, 0, 0, DarkenCompositeOp);

>The error is no pixels defined in cache.

You are trying to composite two images, `combinedMask` and `maskA`. The message says that at least one of the two images has no pixels. Fairly obviously, `combinedMask` has no pixels.

**Reply by 2246sam20** [reply]
Okay. Thanks for the insight.
If possible can you please check the code once for the possible mistakes? And how to resolve them.

Thanks in advance.

**Reply by 2246sam20** [reply]
> > Image combinedMask;
> > combinedMask.composite(maskA, 0, 0, DarkenCompositeOp);
> 
> > The error is no pixels defined in cache.
> 
> You are trying to composite two images, `combinedMask` and `maskA`. The message says that at least one of the two images has no pixels. Fairly obviously, `combinedMask` has no pixels.

OK..
If possible can you please check the code once for possible mistakes? And how to resolve them.

Thanks in advance.

**Reply by snibgo** [reply]
> If possible can you please check the code once for possible mistakes? And how to resolve them.

No. That is too much work.

You are writing a program. You need to check that it runs without errors, and produces the expected results under all possible conditions. You can insert statements that write intermediate results to files, and check that each is what you expect.

If you have IM problems that you can't solve, you can show your complete program, with all the input files and output files, and explain what the problem is.

**Comment by snibgo** [comment]
I forgot to show `cols.txt`, the text file I used to test the above. The file is:
```
Yellow
\#f00
```
The colour `#f00` is the same as `Red`.

I don't know why we need a backslash `\` in front of the hash `#`.

EDIT: I have discovered that in magick scripts, when hash `#` is encountered, it and characters to the end of line are ignored (treated as a comment). See MagickWand/script-token.c .

We can avoid this behaviour by escaping the hash with a backslash or putting the entire colour in double quotes `"`, eg `"#f00"`.

In addition, lines that have `@` or colon `:` in the first column are entirely ignored.
